RavenDB-13082 : TimeSeries querying - JavaScript projections

aviv86 opened the pull request:
- https://issues.hibernatingrhinos.com/issue/RavenDB-13082
- https://issues.hibernatingrhinos.com/issue/RavenDB-14305
- https://issues.hibernatingrhinos.com/issue/RavenDB-13084

aviv86 committed:
RavenDB-13082 : TimeSeries querying - initial work on JS Projections

aviv86 committed:
RavenDB-13084 : TimeSeries client API - initial work on JS projections

aviv86 committed:
RavenDB-13082 : JS projections - add the query root aliases to script runner in order to support query aliases usage inside timeseries query text

aviv86 committed:
RavenDB-13082 : add more tests

aviv86 committed:
RavenDB-14305 : TimeSeries Querying - provide the user with a way to do arbitrary computation using declare functions - basic tests

aviv86 committed:
RavenDB-13082 : JavaScript projections - allow to call the timeseries declared functions as if it was a JS function, and to process its output

aviv86 committed:
RavenDB-13084 : Linq JS Projections : avoid translation to 'timeseries()' JS call - translate timeserires expressions directly to timeseries declared function

aviv86 committed:
RavenDB-13082 : remove 'timeseries()' javascript clr function (use 'declare timeseries' instead) and revert changes in QueryKey

aviv86 committed:
RavenDB-13082 : minor fixes

aviv86 committed:
RavenDB-14305 : arbitrary computation on timeseries queries via linq

aviv86 committed:
RavenDB-13082 : test querying multiple timeseries and querying timeseries with multiple values

aviv86 committed:
RavenDB-13084 : add 'TimeSeriesSupport' extension to Lambda2JS, in order to support linq JS projections that have nested timeseries call expressions. Also, add parameter expressions handling to TimeSeriesQueryVisitor

aviv86 committed:
RavenDB-13084 : WrappedConstantSupport - add support for Call expressions when Object is a wrapped constant expression (e.g. baseline.AddMonths(2))

aviv86 committed:
RavenDB-13084 : TimeSeriesQueryVisitor - better Parameters handling

aviv86 committed:
RavenDB-13084 : ScriptRunner.InvokeTimeSeriesFunction - take the Id of the document to operate on from the first argument (it can be a different document than the 'original' doc)

aviv86 committed:
RavenDB-13084 : fix failing tests due to changes made in WrappedConstantSupport extension :
put all the 'special methods' extensions (LoadSupport, MetadataSupport, CompareExchangeSupport, CounterSupport) before WrappedConstantSupport extension in the extensions array, in order to avoid WrappedConstantSupport mistakenly recognizing expressions like 'session.Load<Company>(o.Company)' as Call expressions on wrapped constant Object

aviv86 committed:
RavenDB-13084 : LinqMethodsSupport : support OrderBy when the argument is an array index expression (e.g. 'tsResults.OrderBy(entry => entry.Values[0])')

aviv86 committed:
RavenDB-13082 : TimeSeriesRetriever : fix NRE in GetParameterIndex()

aviv86 committed:
RavenDB-13082 : TimeSeries Queries : allow to filter on 'Value' property of TimeSeriesEntry, and to do JS computations on Value of entries of timeseries query result

ayende commented on src/Raven.Server/Documents/Patch/ScriptRunner.cs:
```diff
@@ -748,6 +761,82 @@ private JsValue IncrementCounter(JsValue self, JsValue[] args)
+                // todo aviv : write meaningful error messages
+
+                AssertValidDatabaseContext();
+                if (args.Length < 1)
```
```suggestion
                if (args.Length != 1)
```

ayende commented on src/Raven.Server/Documents/Patch/ScriptRunner.cs:
```diff
@@ -748,6 +761,82 @@ private JsValue IncrementCounter(JsValue self, JsValue[] args)
+
+                AssertValidDatabaseContext();
+                if (args.Length < 1)
+                    throw new InvalidOperationException($"must be called with exactly 1 string argument");
```
Is this true? Can't we call time series function with more than one arg? 
What is we want to pass additional parameters?

aviv86 replied:
no, this is wrong
this error message was copy & pasted from somewhere else
need to write proper error messages

ayende commented on src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:
```diff
@@ -704,12 +702,51 @@ public override void GenerateScript(ScriptRunner runner)
                 }
             }
+
+            private static string GenerateTimeSeriesInvocationCall(string name, List<QueryExpression> parameters)
```
why do we need this? Why not inject the call directly as CLR func?
Can we get the name that the function was called under?

aviv86 replied:
done

ayende commented on test/SlowTests/Client/TimeSeries/Query/TimeSeriesJavascriptProjections.cs:
```diff
@@ -0,0 +1,2008 @@
+@"declare timeseries out(d) 
+{
+    from d.Stocks between $start and $end
+    where Tag != 'tags/2'
```
How can I pass this as a param?
So for `Series`, it will be `tags/2` but for `Series2` it will be `tags/3` ?

aviv86 replied:
```
declare timeseries out(d, t) 
{
    from d.Stocks between $start and $end
    where Tag != t
}
from People as p
where p.Age > 49
load p.WorksAt as Company
select {
    Series: out(p, 'tags/2'),
    Series2: out(Company, 'tags/3')
}
```

added test

ayende commented on test/SlowTests/Client/TimeSeries/Query/TimeSeriesJavascriptProjections.cs:
```diff
@@ -0,0 +1,2008 @@
+select {
+    Heartrate: out('Heartrate'),
+    Stocks: out('Stocks')
+}")
```
I like this format

ayende commented on test/SlowTests/Client/TimeSeries/Query/TimeSeriesJavascriptProjections.cs:
```diff
@@ -0,0 +1,2008 @@
+    var allTsNames = d['@metadata']['@timeseries'];
+    for (var i = 0; i < allTsNames.length; i++){
+        var name = allTsNames[i];
+        result[name] = ts(name);
```
How can I pass different ranges here?

aviv86 replied:
```
declare timeseries ts(name, start, end) 
{
    from name between start and end
    where Tag != 'tags/2'
}
```
and then we can pass a diffrent start & end for each call 

added test

ayende commented on test/SlowTests/Client/TimeSeries/Query/TimeSeriesJavascriptProjections.cs:
```diff
@@ -0,0 +1,2008 @@
+                                            Avg = g.Average(),
+                                            Max = g.Max()
+                                        })
+                                        .ToList(),
```
Is the ToList needed? 

I really like what you have done here, the API is beautiful

aviv86 committed:
RavenDB-13082 : fix failing test ShouldUseHasOwnProperty() - WrappedConstantSupport should avoid trying to get value from call-expressions on wrapped constants if the call expression has 'Parameter' arguments

aviv86 committed:
RavenDB-13084 : TimeSeriesQueryVisitor - support parameter expressions inside where clause

aviv86 committed:
RavenDB-13082 : add tests

aviv86 committed:
RavenDB-13082 : support calling timeseries declaration as if it was a JS function :
remove GenerateTimeSeriesInvocationCall() - instead of replacing function bodies with invocation scripts, directly inject the time series calls as CLR functions

aviv86 committed:
RavenDB-13082 : cleanup

ppekrol commented on src/Raven.Client/Documents/Queries/TimeSeries/TimeSeriesWhereClausVisitor.cs:
```diff
@@ -1,28 +1,32 @@
 namespace Raven.Client.Documents.Queries.TimeSeries
 {
-    internal class TimeSeriesWhereClauseModifier<T> : ExpressionVisitor
+    internal class TimeSeriesWhereClausVisitor<T> : ExpressionVisitor
```
typo: TimeSeriesWhereClauseVisitor

ppekrol commented on src/Raven.Server/Documents/Queries/Results/TimeSeriesRetriever.cs:
```diff
@@ -125,9 +125,10 @@ BlittableJsonReaderObject GetRawValues()
+                        [nameof(TimeSeriesEntry.Tag)] = singleResult.Tag.ToString(),
+                        [nameof(TimeSeriesEntry.Timestamp)] = singleResult.Timestamp,
+                        [nameof(TimeSeriesEntry.Values)] = vals,
+                        [nameof(TimeSeriesEntry.Value)] = singleResult.Values.Span[0]
```
Do we need to return this? Looks like redundant data?

ppekrol commented on src/Raven.Client/Util/JavascriptConversionExtensions.cs:
```diff
@@ -2409,6 +2435,32 @@ private static bool CanConvert(Expression expression, DocumentConventions conven
             }
         }
 
+        public class TimeSeriesSupport<T> : JavascriptConversionExtension
```
internal?

ppekrol commented on src/Raven.Client/Documents/Session/Tokens/DeclareToken.cs:
```diff
@@ -5,39 +5,50 @@ namespace Raven.Client.Documents.Session.Tokens
 {
     public class DeclareToken : QueryToken
     {
+        public enum DeclarationType
```
Let's remove that and create maybe two methods.

`CreateFunction` and `CreateTimeSeries`

aviv86 committed:
RavenDB-13082 : address PR comments
